library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,568 @@
+  # especially with storing and initializing widgets it can be quite complex. Consider e.g.
+  # instancing of InputField like:
+  #
+  # ```
+  #   widget = InputField.new(
+  #     label: _("My label"),
+  #     help: _("blablbalba" \
+  #       "blablabla" \
+  #       "blablabla"
+  #     ),
+  #     init: Proc.new do |widget|
+  #       ...
+  #     end,
+  #     store: Proc.new do |widget, event|
+  #       ...
+  #     end,
+  #     validate: Proc.new do |widget, event|
+  #       ....
+  #     end
+  #   )
+  # ```
+  #
+  # For this example current subclassing approach looks like
+  # ```
+  #   class MyWidget < CWM::InputField
+  #     def label
+  #       _("My label")
+  #     end
+  #
+  #     def help
+  #       _("blablbalba" \
+  #         "blablabla" \
+  #         "blablabla"
+  #     end
+  #
+  #     def init
+  #       ...
+  #     end
+  #
+  #     def store
+  #       ...
+  #     end
+  #
+  #     def validate
+  #       ....
+  #     end
+  #   end
+  #  # and usage in dialog
+  #  widget = MyWidget.new
+  # ```
+  #
+  class AbstractWidget
+    include Yast::UIShortcuts
+    include Yast::I18n
+
+    # @return [String] id used for widget
+    attr_writer :widget_id
+    attr_writer :handle_all_events
+
+    # specify if widget handle all raised events or only its own
+    # By default only own values are handled
+    def handle_all_events
+      @handle_all_events.nil? ? false : @handle_all_events
+    end
+
+    def widget_id
+      @widget_id || self.class.to_s
+    end
+
+    # defines widget type for CWM usage
+    def self.widget_type=(val)
+      define_method(:widget_type) { val }
+    end
+
+    # generates description for CWM with widget. Description is auto filled from defined methods.
+    #
+    # methods used to generate description:
+    #
+    # - `#help` [String] to get translated help text for widget
+    # - `#label` [String] to get translated label text for widget
+    # - `#opt` [Array<Symbol>] to get options passed to widget like `[:hstretch, :vstretch]`
+    # - `#validate` [Boolean ()] validate widget value. Returns false if validation failed.
+    # - `#init` [nil ()] initialize widget like e.g. set initial value
+    # - `#handle` [Symbol,nil (String?)] handle widget changed value or press.
+    #   return value is usually nil, returning symbol can be used to send different event.
+    #   It support varient with parameter specified and without. If parameter is specified,
+    #   gets event Hash including "ID" key with value that specify widget ID that cause event.
+    #   It is mainly useful if handle_all_events is set to true to distinguish which event caused it.
+    # - `#store` [nil ()] store widget value after user confirm dialog
+    # - `#cleanup` [nil ()] cleanup after widget is destroyed
+    # @raise [RuntimeError] if required method is not implemented or widget id not set.
+    def description
```

Comment:
API: In UI context I understand a "description" as a "long label". Should we rename this to `definition` or even `cwm_definition` ? Or mimic its hash key and name it  `widget_descr`?